ignite/metrics/psnr.py

```diff
@@ -0,0 +1,118 @@
+    The output of the engine's ``process_function`` needs to be in format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        psnr = PSNR()
+        psnr.attach(engine, "psnr")
+        # ...
+        state = engine.run(data)
+        print(f"PSNR: {state.metrics['psnr']}")
+
+    .. versionadded:: 0.5.0
+    """
+
+    def __init__(
+        self,
+        data_range: Optional[Union[int, float]] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ):
+        super().__init__(output_transform=output_transform, device=device)
+        self.data_range = self.__data_range = data_range
+
+    def _check_shape_dtype(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred, y = output
+        if y_pred.dtype != y.dtype:
+            raise TypeError(
+                f"Expected y_pred and y to have the same data type. Got y_pred: {y_pred.dtype} and y: {y.dtype}."
+            )
+
+        if y_pred.shape != y.shape:
+            raise ValueError(
+                f"Expected y_pred and y to have the same shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_squared_error = 0  # type: Union[int, torch.Tensor]
+        self._num_elements = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        self._check_shape_dtype(output)
+        y_pred, y = output[0].detach(), output[1].detach()
+
+        if self.__data_range is None:
+            dmin, dmax = _dtype_range[y.dtype]
+            true_min, true_max = y.min(), y.max()
+            if true_max > dmax or true_min < dmin:
+                raise ValueError(
+                    "y has intensity values outside the range expected "
+                    "for its data type. Please manually specify the data_range."
+                )
+            if true_min >= 0:
+                # most common case (255 for uint8, 1 for float)
+                self.data_range = dmax
+            else:
+                self.data_range = dmax - dmin
+
+        self._sum_of_squared_error += (
+            torch.pow(y_pred.double() - y.view_as(y_pred).double(), 2).sum().to(device=self._device)
+        )
+        self._num_elements += y.numel()
+
+    @sync_all_reduce("_sum_of_squared_error", "_num_elements")
+    def compute(self) -> Union[float, torch.Tensor]:
+        if self._num_elements == 0:
+            raise NotComputableError("PSNR must have at least one example before it can be computed.")
+        return 10.0 * torch.log10(
+            self.data_range ** 2 / (self._sum_of_squared_error / self._num_elements)  # type: ignore[operator, arg-type]
```

Comment:
I'm still not very certain about this implementation. Ultimately, we would like to compute average PSNR as 
```python
avg_psnr = sum_on_images(psnr[image]) =

= 10 * log10 ( prod_on_images( data_range[image] / mse[image] ) ) = 
= 10 * log10 ( prod_on_batch( prod_images_in_batch( data_range[image] / mse[image] ) ) = 
= 10 * log10 ( prod_on_batch( accumulated_pow10_pnsr[batch] ) ) 
```

What do you think ?

Reply:
The thing is we allow any shape for the metrics, so we can't assume the first dim is batch_size and cannot compute mean for the rest of dims.

Reply:
Well, for the images I've never seen data with shapes like `(..., B, ...)`. Otherwise, we can put this as a requirement which is rather natural.

Reply:
I don't have an example, I usually have NCHW images. Since skimage allows any shape, I try to adopt that.
[TensorFlow](https://www.tensorflow.org/api_docs/python/tf/image/psnr) has some restrictions tho.
Should we allow only 4 dim use case?
Is there 2/3 dim use case I am not aware of?
Also, this metrics can apply to audio domain as well (I am completely new to that domain).

Reply:
Well, I think a support for case starting from 2D would be great ! Let's see what is possible to do

Reply:
Supported now